Public/Src/Cache/ContentStore/Distributed/NuCache/ClusterState.cs

```diff
@@ -130,5 +131,25 @@ public bool TryResolveMachineId(MachineLocation machineLocation, out MachineId m
         {
             return _idByLocationMap.TryGetValue(machineLocation, out machineId);
         }
+
+        /// <summary>
+        /// Gets a random locations excluding the specified location. Returns default if operation is not possible.
+        /// </summary>
+        public MachineLocation GetRandomMachineLocation(MachineLocation except)
+        {
+            if (_idByLocationMap.Keys.Any(location => !location.Equals(except)))
+            {
+                MachineLocation location;
+                do
+                {
+                    location = _idByLocationMap.Keys.ElementAt(ThreadSafeRandom.Generator.Next(_idByLocationMap.Keys.Count));

```

Comment:
YOu may consider getting all the keys into a local because the Keys property is very expensive:

It creates a copy of all the keys and it acquires all the locks.